File: src/AutoRest.SdkExplorer/Model/Schema/SchemaObject.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;
using AutoRest.SdkExplorer.Model.Code;
using YamlDotNet.Serialization;

namespace AutoRest.SdkExplorer.Model.Schema
{
    public class SchemaObject : SchemaBase
    {
        public const string SCHEMA_TYPE = "OBJECT_SCHEMA";

        public List<SchemaProperty> Properties { get; set; } = new List<SchemaProperty>();
        public TypeDesc? InheritFrom { get; set; }
        // TODO: discriminator support
        public List<TypeDesc> InheritBy { get; set; } = new List<TypeDesc>();
        public SchemaMethod? InitializationConstructor { get; set; }
        public SchemaMethod? SerializationConstructor { get; set; }
        /// <summary>
        /// StaticCreateMethod is also treated as one kind of constructor
        /// </summary>
        public SchemaMethod? StaticCreateMethod { get; set; }
        public string Description { get; set; } = string.Empty;
        public bool IsStruct { get; set; }
        public bool IsEnum { get; set; }
        public List<SchemaEnumValue> EnumValues { get; set; } = new List<SchemaEnumValue>();
        public string DiscriminatorKey { get; set; } = string.Empty;
        public bool IsDiscriminatorBase { get; set; } = false;
        public SchemaProperty? DiscriminatorProperty { get; set; }

        /// <summary>
        /// Include properties from inherit
        /// </summary>
        /// <returns></returns>
        public IEnumerable<SchemaProperty> GetAllProperties(SchemaStore? store = null)
        {
            foreach (var sp in this.Properties)
                yield return sp;
            var parent = this.InheritFrom?.GetSchema(store);
            if (parent != null)
            {
                if (parent is not SchemaObject so)
                    throw new InvalidOperationException("Object Schema is expected for parent object: " + this.SchemaKey);
                foreach (var r in so.GetAllProperties(store))
                    yield return r;
            }
        }

        private SchemaObject()
            : base("OnlyForYamlDeserializor", SCHEMA_TYPE)
        {
        }

        [JsonConstructor]
        public SchemaObject(string schemaKey)
            : base(schemaKey, SCHEMA_TYPE)
        {
        }

        /// <summary>
        /// Static Creation method is also treated as one kind of constructor
        /// </summary>
        /// <returns></returns>
        public SchemaMethod? GetDefaultConstructor() => this.SerializationConstructor ?? this.InitializationConstructor ?? this.StaticCreateMethod ?? null;

        public bool IsDefaultConstructorParameter(string parameterSerializerName)
        {
            var ctor = this.GetDefaultConstructor();
            if (ctor == null)
                return false;
            else
                return ctor.HasParameter(parameterSerializerName);
        }

    }
}

File: src/AutoRest.SdkExplorer.TypeScript/src/Model/Schema/SchemaObject.ts
import { TypeDesc } from "../Code/TypeDesc";
import { SchemaBase } from "./SchemaBase";
import { SchemaEnumValue } from "./SchemaEnumValue";
import { SchemaMethod } from "./SchemaMethod";
import { SchemaProperty } from "./SchemaProperty";

export class SchemaObject extends SchemaBase implements AutoRest.SdkExplorer.Interface.SchemaObject {
    properties: SchemaProperty[];
    inheritFrom?: TypeDesc;
    inheritBy: TypeDesc[];
    initializationConstructor?: SchemaMethod;
    serializationConstructor?: SchemaMethod;
    staticCreateMethod?: SchemaMethod;
    description: string;
    isStruct: boolean;
    isEnum: boolean;
    enumValues: SchemaEnumValue[];
    discriminatorKey?: string;
    isDiscriminatorBase: boolean;
    discriminatorProperty?: SchemaProperty;

    constructor(data: AutoRest.SdkExplorer.Interface.SchemaObject) {
        super(data);
        this.properties = data.properties?.map(d => new SchemaProperty(d)) ?? []
        this.inheritFrom = data.inheritFrom ? new TypeDesc(data.inheritFrom) : undefined;
        this.inheritBy = data.inheritBy?.map(d => new TypeDesc(d)) ?? [];
        this.initializationConstructor = data.initializationConstructor ? new SchemaMethod(data.initializationConstructor) : undefined;
        this.serializationConstructor = data.serializationConstructor ? new SchemaMethod(data.serializationConstructor) : undefined;
        this.staticCreateMethod = data.staticCreateMethod ? new SchemaMethod(data.staticCreateMethod) : undefined;
        this.description = data.description ?? "";
        this.isStruct = data.isStruct!;
        this.isEnum = data.isEnum!;
        this.enumValues = data.enumValues?.map(d => new SchemaEnumValue(d)) ?? [];
        this.discriminatorKey = data.discriminatorKey;
        this.isDiscriminatorBase = data.isDiscriminatorBase ?? false;
        this.discriminatorProperty = data.discriminatorProperty ? new SchemaProperty(data.discriminatorProperty) : undefined;
    }

    public get defaultCreateMethod(): SchemaMethod | undefined {
        return this.serializationConstructor ?? this.initializationConstructor ?? this.staticCreateMethod ?? undefined;
    }
    public get isDefaultCreateMethodStatic(): boolean {
        return this.staticCreateMethod !== undefined && this.defaultCreateMethod === this.staticCreateMethod;
    }
    public get extraProperties(): SchemaProperty[] | undefined {
        let ctorParamNames = this.defaultCreateMethod?.methodParameters?.map(mp => mp.name?.toLowerCase()) ?? [];
        let r: SchemaProperty[] = this.properties ?? [];
        if (this.inheritFrom) {
            r = r.concat(this.inheritFrom.schemaObject?.properties ?? []);
        }
        return r.filter(p => !ctorParamNames.includes(p.name?.toLowerCase()));
    }
}
